cpp/src/aztec/polynomials/polynomial.cpp

```diff
@@ -48,112 +37,63 @@ Polynomial<Fr>::Polynomial(std::string const& filename)
-    , size_(initial_size_)
-    , page_size_(DEFAULT_SIZE_HINT)
-    , max_size_(0)
-    , allocated_pages_(0)
+Polynomial<Fr>::Polynomial(const size_t size_)
+    : coefficients_(nullptr)
+    , size_(size_)
+    , mapped_(false)
 {
-    ASSERT(page_size_ != 0);
-    size_t target_max_size = std::max(initial_size_, initial_max_size_hint + DEFAULT_PAGE_SPILL);
-    if (target_max_size > 0) {
-
-        max_size_ = target_max_size;
-        coefficients_ = (Fr*)(aligned_alloc(32, sizeof(Fr) * target_max_size));
+    if (capacity() > 0) {
+        coefficients_ = (Fr*)(aligned_alloc(32, sizeof(Fr) * capacity()));
     }
-    memset(static_cast<void*>(coefficients_), 0, sizeof(Fr) * max_size_);
+    memset(static_cast<void*>(coefficients_), 0, sizeof(Fr) * capacity());
 }
 
 template <typename Fr>
-Polynomial<Fr>::Polynomial(const Polynomial<Fr>& other, const size_t target_max_size)
-    : mapped_(false)
-    , initial_size_(other.initial_size_)
-    , size_(other.size_)
-    , page_size_(other.page_size_)
-    , max_size_(std::max(clamp(target_max_size, page_size_ + DEFAULT_PAGE_SPILL), other.max_size_))
-    , allocated_pages_(max_size_ / page_size_)
+Polynomial<Fr>::Polynomial(const Polynomial<Fr>& other, const size_t target_size)
+    : size_(std::max(target_size, other.size()))
+    , mapped_(false)
 {
-    ASSERT(page_size_ != 0);
-    ASSERT(max_size_ >= size_);
-
-    coefficients_ = (Fr*)(aligned_alloc(32, sizeof(Fr) * max_size_));
+    coefficients_ = (Fr*)(aligned_alloc(32, sizeof(Fr) * capacity()));
 
     if (other.coefficients_ != nullptr) {
-        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * size_);
+        memcpy(static_cast<void*>(coefficients_), static_cast<void*>(other.coefficients_), sizeof(Fr) * other.size_);
     }
-    zero_memory(max_size_);
+    zero_memory(other.size_, size_);
 }
+
 template <typename Fr>
 Polynomial<Fr>::Polynomial(Polynomial<Fr>&& other) noexcept
-    : mapped_(other.mapped_)
-    , coefficients_(other.coefficients_)
-    , initial_size_(other.initial_size_)
-    , size_(other.size_)
-    , page_size_(other.page_size_)
-    , max_size_(other.max_size_)
-    , allocated_pages_(other.allocated_pages_)
-{
-    ASSERT(page_size_ != 0);
-    other.coefficients_ = 0;
-
-    // Clear other, so we can detect use on free after poly's are put in cache
-    other.clear();
-}
+    : coefficients_(std::exchange(other.coefficients_, nullptr))
+    , size_(std::exchange(other.size_, 0))
+    , mapped_(std::exchange(other.mapped_, false))
+{}
 
 template <typename Fr>
-Polynomial<Fr>::Polynomial(Fr* buf, const size_t initial_size_)
-    : mapped_(false)
-    , coefficients_(buf)
-    , initial_size_(initial_size_)
-    , size_(initial_size_)
-    , page_size_(DEFAULT_SIZE_HINT)
-    , max_size_(initial_size_)
-    , allocated_pages_(0)
+Polynomial<Fr>::Polynomial(Fr* buf, const size_t size_)
+    : coefficients_(buf)
+    , size_(size_)
+    , mapped_(false)
 {}
```

Comment:
This function is used for reading polys from a `mmap`. It may be possible to get rid of it at some point but it's used at various points in our testing suite, for example, to reuse a proving key that has been written to file.